### ISSUE TITLE
stax neural network initializers default to 32bit floats even in 64 bit mode

### ISSUE THREAD

**Issue by lumip** [google/jax#2003]
I need/want to try out my code with 64 bit wide floats. My code uses neural networks setup with stax, the initializers for which default to 32 bit floats (and there is no useable API to change this in user facing code).

Using the thus initialized (`float32`) parameters (and `float64` data batches) in an update `fori_loop` (taking the gradients with `value_and_grad` and updating parameters with any optimizer) results in either one of two faulty behaviors (depending on the exact nature of computation):

1. the gradients come out as `float32`, thus thwarting my intention of computing with `float64` values
2. the gradients come out as `float64`, which results in a type-mismatch exception thrown by `fori_loop` since the initial value argument passed to it was of type `float32`

**Comment by hawkinsp** [comment]
`stax` is intentionally a somewhat minimal example of a neural network library that is perhaps more intended for expository purposes than anything else. Its goal is to be simple more than it is to be feature complete.

There are a number of other neural network libraries built on top of JAX that may be more useful and more feature complete (examples include [flax](https://github.com/google-research/flax/tree/prerelease), [trax](https://github.com/google/trax), [Jaxnet](https://github.com/JuliusKunze/jaxnet), ... and there are more! A thousand flowers seem to be blooming right now, which is great!) I don't know if any of them in particular support 64-bit weights, but it seems like a reasonable thing to want.

That said, `stax` is very minimal, only a few hundred lines of code. So if you are mostly happy with stax, I'd actually just suggest forking it into your repository and making whatever changes you like. Or feel free to send a PR allowing more dtype flexibility. Up to you. It might also be possible to get the type behavior you want by passing in a custom initializer.

**Comment by lumip** [comment]
Would you consider a PR that just changes all default `dtypes` in `jax/nn/initializers.py` to `np.float64`? With default jax behavior (i.e., 32 bit mode) those will be cast to `np.float32` anyways so nothing changes there but in 64 bit mode (i.e., when `jax_enable_x64` was explicitly set to `True`) they would then always be `float64`.

**Comment by hawkinsp** [comment]
I don't think that would be a good idea, because our plan is to make x64 mode the default sometime soon. The proposed change would mean that the default is simply 64-bit weights, which seems like a strange default for neural network training. I think the best plan is to make the dtype user-selectable somehow.
